common/chains/client/multi_node_client.go

```diff
@@ -0,0 +1,324 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+)
+
+var (
+	// PromMultiNodeClientRPCNodeStates reports current RPC node state
+	PromMultiNodeClientRPCNodeStates = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "client_rpc_node_states",
+		Help: "The number of RPC nodes currently in the given state for the given chain",
+	}, []string{"network", "chainId", "state"})
+)
+
+const (
+	NodeSelectionMode_HighestHead     = "HighestHead"
+	NodeSelectionMode_RoundRobin      = "RoundRobin"
+	NodeSelectionMode_TotalDifficulty = "TotalDifficulty"
+	NodeSelectionMode_PriorityLevel   = "PriorityLevel"
+)
+
+type NodeSelector[
+	CHAIN_ID types.ID,
+	HEAD Head,
+	RPC_CLIENT NodeClient[CHAIN_ID, HEAD],
+] interface {
+	// Select returns a Node, or nil if none can be selected.
+	// Implementation must be thread-safe.
+	Select() Node[CHAIN_ID, HEAD, RPC_CLIENT]
+	// Name returns the strategy name, e.g. "HighestHead" or "RoundRobin"
+	Name() string
+}
+
+type MultiNodeClient[
+	CHAIN_ID types.ID,
+	HEAD Head,
+	RPC_CLIENT NodeClient[CHAIN_ID, HEAD],
+	TX any,
+] interface {
+	Dial(context.Context) error
+	Close() error
+	NodeStates() map[string]string
+	SelectNode() Node[CHAIN_ID, HEAD, RPC_CLIENT]
+	NodesAsSendOnlys() []SendOnlyNode[CHAIN_ID, RPC_CLIENT]
+	WrapSendOnlyTransaction(ctx context.Context, lggr logger.Logger, tx TX, n SendOnlyNode[CHAIN_ID, RPC_CLIENT],
+		f func(ctx context.Context, lggr logger.Logger, tx TX, n SendOnlyNode[CHAIN_ID, RPC_CLIENT]))
+
+	runLoop()
```

Comment:
Why does runLoop need to be part of this interface? Seems like it is called only from within the struct.